core/server/src/test/java/alluxio/worker/netty/DataServerBlockReadHandlerTest.java

```diff
@@ -0,0 +1,304 @@
+      Object response = waitForOneResponse();
+      DataBuffer buffer = checkReadResponse(response, Protocol.Status.Code.OK);
+      if (buffer == null) {
+        eof = true;
+        break;
+      }
+      buffer.release();
+      maxIterations--;
+      Assert.assertTrue(mChannel.isOpen());
+    }
+    Assert.assertTrue(eof);
+  }
+
+  @Test
+  public void readFailure() throws Exception {
+    long fileSize = PACKET_SIZE * 10 + 1;
+    populateInputFile(0, 0, fileSize - 1);
+    mBlockReader.close();
+    mChannel.writeInbound(buildReadRequest(0, fileSize));
+    Object response = waitForOneResponse();
+    checkReadResponse(response, Protocol.Status.Code.INTERNAL);
+    waitForChannelClose();
+    Assert.assertTrue(!mChannel.isOpen());
+  }
+
+  /**
+   * Populates the input file, also computes the checksum for part of the file.
+   *
+   * @param length the length of the file
+   * @param start the start position to compute the checksum
+   * @param end the last position to compute the checksum
+   * @throws Exception if it fails to populate the input file
+   */
+  private void populateInputFile(long length, long start, long end) throws Exception {
+    mChecksum = 0;
+    File file = mTestFolder.newFile();
+    long pos = 0;
+    if (length > 0) {
+      FileOutputStream fileOutputStream = new FileOutputStream(file);
+      while (length > 0) {
+        byte[] buffer = new byte[(int) Math.min(length, Constants.MB)];
+        mRandom.nextBytes(buffer);
+        for (int i = 0; i < buffer.length; i++) {
+          if (pos >= start && pos <= end) {
+            mChecksum += BufferUtils.byteToInt(buffer[i]);
+          }
+          pos++;
+        }
+        fileOutputStream.write(buffer);
+        length -= buffer.length;
+      }
+      fileOutputStream.close();
+    }
+
+    mFile = file.getPath();
+    mBlockReader = new LocalFileBlockReader(mFile);
+    PowerMockito
+        .when(mBlockWorker.readBlockRemote(Mockito.anyLong(), Mockito.anyLong(), Mockito.anyLong()))
+        .thenReturn(mBlockReader);
+  }
+
+  /**
+   * Builds a read request.
+   *
+   * @param offset the offset
+   * @param len the length to read
+   * @return the proto message
+   */
+  private RPCProtoMessage buildReadRequest(long offset, long len) {
+    Protocol.ReadRequest readRequest =
+        Protocol.ReadRequest.newBuilder().setId(mBlockId).setOffset(offset).setLength(len)
+            .setLockId(1L).setType(Protocol.RequestType.ALLUXIO_BLOCK).build();
+    return new RPCProtoMessage(readRequest, null);
+  }
+
+  /**
+   * Checks all the read responses.
+   */
+  private void checkAllReadResponses() {
+    int timeRemaining = Constants.MINUTE_MS;
+    boolean eof = false;
+    long checksumActual = 0;
+    while (!eof && timeRemaining > 0) {
+      Object readResponse = null;
+      while (readResponse == null && timeRemaining > 0) {
+        readResponse = mChannel.readOutbound();
+        CommonUtils.sleepMs(10);
+        timeRemaining -= 10;
```

Comment:
it doesn't need to be accurate since it is unittest.  I've changed it currentTimeMillis()